### PR TITLE
Add oriented bounding box (OBB)

### PR DESCRIPTION
**Pull request by aprokop**
[Currently contains #1128]

This PR adds oriented bounding box geometry. This geometry consists of an orthonormal transformation matrix and a AABB in the new coordinate system. We use SVD to solve the eigenvalue problem. 

There are some nuances with the implementation. Due to using floating point operations, there could be false positives and negatives for intersection tests, as well as an error in the distance metric. The only way to get rid of that is to expand the box by some amount in each direction. However, this makes the testing pretty hard to do. Right now, in this draft form, the box is not expanded so that we can have discussion. Because of the false stuff, using OBB for MST won't work, it will break down (my understanding is that that distance to the OBB may be larger than the distance to the contained primitives, which results in problems).

I'd like to have a discussion about how much to expand a box by, as well as how to deal with that in the testing.

### CODE REVIEW

**Comment by Rombur** [comment]
This is great. I will use this in adamantine.

File: test/tstOBB.cpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    BOOST_TEST(!intersects(Point{-0.5f, 0.49f}, obb));
+    BOOST_TEST(intersects(Point{-0.25f, 0.74f}, obb));
+    BOOST_TEST(intersects(Point{0.25f, -0.24f}, obb));
+    BOOST_TEST(!intersects(Point{0.5f, 0.51f}, obb));
+    BOOST_TEST(intersects(Point{0.75f, 0.24f}, obb));
+  }
+  {
+    // OBB is not necessarily the tightest box around the points. Here's an
+    // example where the covariance matrix is proportional to identity, and
+    // we simply choose the default axes, i.e.,
```
Comment (Rombur) [inline test/tstOBB.cpp:111]: Do you know why that's the case?

Comment (aprokop) [reply test/tstOBB.cpp:111]: My understanding that this is due to the sampling bias. Corners of a polytope may not represent it sufficiently well from the probabilistic sampling point of view. See also Sections 3.3.1 and 3.3.2 of the Gottschalk's [dissertation](https://gamma.cs.unc.edu/users/gottschalk/main.pdf).

**Comment by aprokop** [comment]
> This is great. I will use this in adamantine.

What will you use it for?

**Comment by Rombur** [comment]
When we deposit material in adamantine, we need to activate new cells in the mesh. To know which cells to activate, we compute the intersection of the mesh with a box that represents the material deposition. In general this box is not axis-aligned, so we split it in a bunch of small AABB that we give to ArborX. With OBB, I won't need to split the box is so many small ones.  I will still need to split large boxes in smaller ones to have a smooth material deposition but I won't need to care about the angle of the boxes anymore.

**Comment by aprokop** [comment]
> When we deposit material in adamantine, we need to activate new cells in the mesh. To know which cells to activate, we compute the intersection of the mesh with a box that represents the material deposition. In general this box is not axis-aligned, so we split it in a bunch of small AABB that we give to ArborX. With OBB, I won't need to split the box is so many small ones. I will still need to split large boxes in smaller ones to have a smooth material deposition but I won't need to care about the angle of the boxes anymore.

Have you tried KDOP?

**Comment by Rombur** [comment]
> Have you tried KDOP?

The problem with KDOP is that the box can have any orientation. If I have to split the box because it is not aligned with the faces of the KDOP, I may as well keep doing what I am doing right now.

> Also, I wonder what exact intersections are you interested in? In general, intersecting box with a obb could be challenging and is not currently implemented.

That's too bad. That's the one I care about.

**Comment by aprokop** [comment]
> That's too bad. That's the one I care about.

It is certainly doable. We can use axis separating theorem and implement it (there are papers on optimizing it too). Just that I didn't need it at the moment and didn't implement it.

**Comment by dalg24** [comment]
Abandoning this?  Please comment.

**Comment by aprokop** [comment]
> Abandoning this? Please comment.

I have to revision how OBBs are done and computed. It is likely that OBBs will be stored using quaternions. However, it is probable that I would need to switch from SVD based algorithm to DiTO. In addition, hierarchy constructions using OBBs does not follow the same pattern and requires multiple passes.

So, in short, I'm not confident this is the right approach. I'll figure it out in the next few months.

**Pull request closed without merging**